### ISSUE TITLE
Question about how to encode the whole doc-context and store it for future retrieval

### ISSUE THREAD

**Issue by zhiwenyou103** [impira/docquery#52]
Hi! I have a question about how to store the doc.context in this command `print(q, p(question=q, **doc.context))`? Because it took a long time to encode context file for each query, so it would be better to pre-encode the context file and use docquery to retrieve from the encoded context. Thank you!

**Comment by zhiwenyou103** [comment]
> To clarify, do you want to cache it in memory or save it to disk?

The ideal way is to store is to the disk and use that for retrieval to accelerate the retrieval.

**Comment by ankrgyl** [comment]
You should be able to pickle `doc.context` (it's just a bunch of tensors). Can you try that?

**Comment by zhiwenyou103** [comment]
> 

Thank you! I stored the tensor as .pt file. However, it still took a long time to complete the retrieval stage for around 100 questions. May I ask how to store it in the memory? Or do you have any other ideas to accelerate this process. Thank you!

**Comment by ankrgyl** [comment]
Well, the `doc.context` property is a `@cached_property`, so it's automatically cached in memory. Can you give me a sense of how many pages are in the document, how long it took to run the 100 questions, what kind of hardware you're running it on (CPU or GPU + specs), and what your expectations are for performance?

**Comment by ankrgyl** [comment]
I'm not sure that would actually make your workload any faster -- you would still do the same # of forward passes (1 per question/window of the doc) regardless of batching. Is it possible to try running your workload on a GPU?

**Comment by zhiwenyou103** [comment]
> I'm not sure that would actually make your workload any faster -- you would still do the same # of forward passes (1 per question/window of the doc) regardless of batching. Is it possible to try running your workload on a GPU?

Yes. I try it on Colab. It still took about 55 min. The for loop I use is the same as you provided in README.
for q in questions:
    answer = p(question=q, **d, top_k=3) #top_k=3
    data = {}
    data['questions'] = q
    data['answers'] = answer
    all_data.append(data)

There is a warning for running this: `UserWarning: You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset`. So I assume it might be the pipeline problem for running slow.

**Comment by ankrgyl** [comment]
Yes, definitely, the pipeline runs one forward pass at a time. If you'd like, you can try disentangling the pipeline, and running things manually (by calling `.forward()` on the underlying model). You could do this over several batches at once, and see if it improves performance. The pipeline implementation is here: https://github.com/impira/docquery/blob/main/src/docquery/ext/pipeline_document_question_answering.py. You'd want to call `preprocess()` separately, and then mimic `_forward()`, but call `model.forward()` directly with a batch you construct.

**Comment by zhiwenyou103** [comment]
> Yes, definitely, the pipeline runs one forward pass at a time. If you'd like, you can try disentangling the pipeline, and running things manually (by calling `.forward()` on the underlying model). You could do this over several batches at once, and see if it improves performance. The pipeline implementation is here: https://github.com/impira/docquery/blob/main/src/docquery/ext/pipeline_document_question_answering.py. You'd want to call `preprocess()` separately, and then mimic `_forward()`, but call `model.forward()` directly with a batch you construct.

Thank you so much!!! Will try it!
